openedx-unsupported/ecommerce
# [ENT-236] Check data sharing consent status more thoroughly for coupons

This pull request adds more thorough checks of user data sharing consent status when redeeming a coupon. At present, users are redirected to the consent flow as long as the attached enterprise has data sharing consent enabled, without regard to whether a user has already provided consent or not. This pull request ensures that users are only directed to the consent flow when they have not already provided consent (for a course or account) and the enterprise enforces consent either at the account level or at the course level.

**JIRA tickets**: Implements [ENT-236](https://openedx.atlassian.net/browse/ENT-236)

**Dependencies**: None

**Merge deadline**: 27 March 2017

**Testing instructions**:

1. Requires a fully-configured course discovery and ecommerce stack.
2. Add a coupon linked to an Enterprise Customer
3. Ensure that the Enterprise Customer has data sharing consent enabled, enforced at enrollment.
3. In the LMS, either log in with an SSO provider linked to the EnterpriseCustomer and provide consent, or manually create a UserDataSharingConsentAudit object via the shell, linked to the user and enterprise in question, with consent enabled.
4. Download the coupon report for the coupon created earlier, and navigate to the "Offers page" URL in that report.
5. Select the course you want to use from the coupon; note that you're taken all the way through the coupon flow and back to the dashboard without being interrupted, rather than being redirected to the consent page.
6. Unlink your user from the Enterprise Customer; set the Enterprise Customer to have data sharing consent enabled, but "optional".
7. De-enroll your user from the course.
8. Again, browse to the Offers page and click through; again, observe that you're not stopped to provide consent.
